unity/unity_utils.c

```diff
@@ -198,3 +200,216 @@ unity_mono_method_is_generic (MonoMethod* method)
+{
+	// Pretty much mimicing load_cattr_value in reflection.c here now, but without the allocs:
+	switch(type)
+	{
+		case MONO_TYPE_U1:
+		case MONO_TYPE_I1:
+		case MONO_TYPE_BOOLEAN:
+		{
+			*(MonoBoolean*)dest = *(MonoBoolean*)src;
+			return 1;
+		}
+		case MONO_TYPE_CHAR:
+		case MONO_TYPE_U2:
+		case MONO_TYPE_I2:
+		{
+			*(guint16*)dest = *(guint16*)src;
+			return 2;
+		}
+#if SIZEOF_VOID_P == 4
+		case MONO_TYPE_U:
+		case MONO_TYPE_I:
+#endif
+		case MONO_TYPE_R4:
+		case MONO_TYPE_U4:
+		case MONO_TYPE_I4: {
+			*(guint32*)dest = *(guint32*)src;
+			return 4;
+		}
+#if SIZEOF_VOID_P == 8
+		case MONO_TYPE_U: /* error out instead? this should probably not happen */
+		case MONO_TYPE_I:
+#endif
+		case MONO_TYPE_U8:
+		case MONO_TYPE_I8: {
+			*(guint64*)dest = *(guint64*)src;
+			return 8;
+		}
+		case MONO_TYPE_R8: {
+			*(double*)dest = *(double*)src;
+			return 8;
+		}
+		case MONO_TYPE_STRING: {
+			const char* oldSrc = src;
+			// if the first byte is 0xFF, it's a null/empty string
+			UnityMonoMetadataString** str = (UnityMonoMetadataString**)dest;
+			if(*src == (char)0xFF)
+			{
+				*str = NULL;
+				return 1;
+			}
+			
+			*str = (UnityMonoMetadataString*)src;
+			src += mono_metadata_decode_value(src, &src);
+			return src - oldSrc;
+		}
+		case MONO_TYPE_CLASS: {
+			const char* oldSrc = src;
+			guint32 length;
+			char* utf8Str;
+
+			// type references (i.e. typeof(SomeClass)) are stored as string names
+			MonoType** result = (MonoType**)dest;
+			if(*src == (char)0xFF)
+			{
+				*result = NULL;
+				return 1;
+			}
+			
+			length = mono_metadata_decode_value (src, &src);
+			utf8Str = g_memdup (src, length + 1);
+			utf8Str [length] = 0;
+			src += length;
+			*result = mono_reflection_type_from_name (utf8Str, image);
+			if (!*result)
+				g_warning ("Cannot load type '%s'", utf8Str);
+			g_free (utf8Str);
+			return src - oldSrc;
+		}
+		default:
+		{
+			g_error ("Type 0x%02x not handled in custom attr value decoding", type);
+			return 0;
+		}
+	}
+}
+
+void unity_mono_unpack_custom_attr_params(MonoImage* image, MonoCustomAttrEntry* entry, void* output, UnityMonoCustomAttrNamedParameterFunc namedParameterHandler)
+{
+	// Mimicing create_custom_attr in mono/metadata/reflection.c for unpacking the params
+	// Also helpful: http://geekswithblogs.net/simonc/archive/2011/06/03/anatomy-of-a-.net-assembly---custom-attribute-encoding.aspx
+	const char *src = (const char*)entry->data;
+	char *dest = (char*)output;
+	int i;
+	guint32 num_named;
+	
+	MonoMethodSignature *sig = mono_method_signature(entry->ctor);
+	
+	// Skip prolog
+	src += 2;
+	for(i = 0; i < sig->param_count; ++i)
+	{
+		MonoType* paramType = sig->params[i];
+		int type = paramType->type;
+		
+		if (type == MONO_TYPE_VALUETYPE && paramType->data.klass->enumtype)
+			type = mono_class_enum_basetype(paramType->data.klass)->type;
+		
+		src += unpack_custom_attr_value(src, type, dest, image);
```

Comment:
I think it would be nice if unity_mono_unpack_custom_attr_params returned the number of params that failed to be unpacked. So if unpack_custom_attr_value hits one of the errors cases, sig->param_count-i is returned immediately.